examples/parsing/string_interpolation.an

```diff
@@ -0,0 +1,19 @@
+shout x = "${x}!"
+
+bar = "nested interpolations"
+
+print
+  "I can do line breaks${
+    shout "\nAny expressions"
+  }${
+    shout "\nEven ${bar}"
+  }"
+
+// args: --parse
+// expected stdout:
+// (shout = (fn x -> ('++' "" ('++' (cast x) "!"))));
```

Comment:
I'd definitely like to see this simplified to
```rs
(shout = (fn x -> ('++' (cast x) "!")));
```
Without the extra empty string literal if possible. Seems like a check could be added in the desugar parser functions for it. 

Also, although it should be inferred the majority of the time, I believe the `: string` may still be necessary to add in the desugaring after casting. My thinking is given a literal such as `"${foo}"` if we simplify out the empty string literal appendings we are left with `cast foo` as the desugared code which may not always resolve to a string without the extra annotation.

Reply:
Empty leading strings are now pruned in the concatenation, and if that happens a type annotation gets added.